Archives/VolFile.cpp

```diff
@@ -389,103 +382,102 @@ namespace Archives
 	bool VolFile::ReadVolHeader()
 	{
-		int temp;
-
-		// Make sure file is big enough to contain header tag
-		if (m_MappedFileSize < 8) {
-			return false;
-		}
-
-		// Make sure the header is structured properly
-		// -------------------------------------------
-		// -------------------------------------------
+		m_HeaderLength = ReadTag(std::array<char, 4> { 'V', 'O', 'L', ' ' });
 
-		// Check opening header tag ("VOL " tag)
-		// -------------------------------------
-		if ((m_HeaderLength = ReadTag(0, "VOL ")) == -1) {
-			return false;
+		uint32_t volhSize = ReadTag(std::array<char, 4> { 'v', 'o', 'l', 'h' });
+		if (volhSize != 0) {
+			return false; //the size of this section must be 0
 		}
 
-		// Make sure the file is large enough to contain the header
-		if (m_MappedFileSize < m_HeaderLength + 8) {
-			return false;
-		}
+		m_StringTableLength = ReadTag(std::array<char, 4> { 'v', 'o', 'l', 's' });
 
-		// Check next header tag ("volh" tag)
-		// ----------------------------------
-		// Note: the size of this section must be 0
-		if (ReadTag(8, "volh") != 0) {
-			return false;
-		}
-
-		// Check for fileName string table tag ("vols" tag)
-		// ------------------------------------------------
-		if ((m_StringTableLength = ReadTag(16, "vols")) == -1) {
-			return false;
-		}
-
-		m_ActualStringTableLength = *((int*)m_BaseOfFile + 6);	// Store actual length
-		m_StringTable = (char*)m_BaseOfFile + 28;				// Store pointer to table
+		ReadStringTable();
 
 		// Make sure the string table fits in the header (and next header fits too)
 		if (m_HeaderLength < m_StringTableLength + 20) {
 			return false;
 		}
 
-		temp = m_StringTableLength + 24;			// Calculate offset to next section
-		// Check for volume index tag ("voli" tag)
-		// ---------------------------------------
-		// Store the offset to the Index table while it's still calculated
-		m_Index = (IndexEntry*)((char*)m_BaseOfFile + temp + 8);
-		if ((m_IndexTableLength = ReadTag(temp, "voli")) == -1) {
-			return false;
+		m_IndexTableLength = ReadTag(std::array<char, 4> { 'v', 'o', 'l', 'i' });
+		if (m_IndexTableLength > 0) {
+			m_IndexEntries.resize(m_IndexTableLength / sizeof(IndexEntry));
+			archiveFileReader->Read(m_IndexEntries.data(), m_IndexTableLength);
 		}
 
 		// Make sure the index table fits in the header
 		if (m_HeaderLength < m_StringTableLength + m_IndexTableLength + 24) {
 			return false;
 		}
 
-		// Determine the number of index entries
-		// -------------------------------------
 		m_NumberOfIndexEntries = m_IndexTableLength / 14;
+		ReadPackedFileCount();
+
+		return true;
+	}
+
+	void VolFile::ReadStringTable()
+	{
+		uint32_t actualStringTableLength;
+		archiveFileReader->Read(&actualStringTableLength, sizeof(actualStringTableLength));
+
+		std::string charBuffer;
+		charBuffer.resize(actualStringTableLength);
+		archiveFileReader->Read(&charBuffer[0], actualStringTableLength);
+
+		m_StringTable.push_back("");
```

Comment:
Why?

Reply:
I think you are referinng to the empty push_back call. There needs to be a string present to add characters to for when the loop starts.